### PR TITLE
Add JSON API and build project pages from JSON data - close #16, close #178, close #82

### LINKED ISSUE #16: Provide an API
It would be nice if we could provide some information via an API - nothing fancy, just some static JSON generated from the DB.

Either one big JSON blob containing everything, or perhaps a few more distinct sets of data.

I'm thinking:

- List of packages (could be limited to those we have available, or contain metadata including success / build output
- List of versions per package
- Build output

### LINKED ISSUE #82: Stats page
- Total number of packages/versions
- Total number of downloads
- Top 10 packages (number of downloads)
- OS/distro/version breakdown
- Pi model breakdown (I guess just Armv6/Armv7)
- Total build time
- Total time saved
- Number of builds pending by ABI
- Build success %

Suggestions?

### LINKED ISSUE #178: Show "all" as ABI in versions table for pure python wheels
In the versions table on project pages, instead of just listing the lowest ABI (cp34m), it would be better if we were able to deduce that it's pure python and show "all".

Screenshot from 2019-07-02 13-34-37

It may give the impression that we only have a working wheel for Jessie.

### PR DESCRIPTION
**Pull request by bennuttall**
- Adds `/project/<project>/json`
- Adds `/json.html`
- Adds `/statistics.json`
- Changes project page to get data from JSON
- Changes homepage to get data from JSON
- Adds contents to stats page at `/stats.html`
- Closes #16, #178, #82 

Todo:

- ensure HTML pages and JSON files are written out at the right time, and at the right frequency
- create queries for missing stats data
- add usage/examples for `statistics.json`
- update tests

### CODE REVIEW

File: piwheels/master/db.py [hunk 1]
```diff
@@ -60,7 +60,9 @@
 
 
 ProjectVersionsRow = namedtuple('ProjectVersionsRow', (
-    'version', 'skipped', 'builds_succeeded', 'builds_failed'))
+    'version', 'released', 'build_id', 'skip', 'duration', 'status', 'filename',
+    'filesize', 'filehash', 'builder_abi', 'file_abi_tag', 'platform_tag',
+    'apt_dependencies'))
```
Comment (cclauss) [inline piwheels/master/db.py]: OPTIONAL:  Also possible and easier to read and write:
```python
ProjectVersionsRow = namedtuple('ProjectVersionsRow', """
    version released build_id skip duration status filename filesize
    filehash builder_abi file_abi_tag platform_tag apt_dependencies""")
```
    'version', 'skipped', 'builds_succeeded', 'builds_failed'))
    'version', 'released', 'build_id', 'skip', 'duration', 'status', 'filename',
    'filesize', 'filehash', 'builder_abi', 'file_abi_tag', 'platform_tag',
    'apt_dependencies'))

**Comment by bennuttall** [comment]
Could add a stats.json file to populate the homepage

File: piwheels/master/the_scribe.py [hunk 2]
```diff
@@ -312,46 +318,91 @@ def write_project_page(self, package):
...
 
+    def write_json_file(self, package):
+        """
+        (Re)writes the JSON file for the specified package.
+
+        :param str package:
+            The name of the package to write the JSON file for
+        """
+        self.logger.info('writing json file for %s', package)
+        pkg_dir = self.output_path / 'project' / package / 'json'
```
Comment (cclauss) [inline piwheels/master/the_scribe.py]: This line works?!?  I would have thought __pkg_dir = "/".join((self.output_path, 'project', package, 'json'))__

Same for line 347

Comment (bennuttall) [reply piwheels/master/the_scribe.py]: :)

Pathlib overloads the `/` operator to allow this: https://docs.python.org/3/library/pathlib.html

Comment (cclauss) [reply piwheels/master/the_scribe.py]: Super cool!!  Thanks!

Comment (waveform80) [reply piwheels/master/the_scribe.py]: I do like the pathlib syntax; it's also entirely "obvious" to the reader what it's doing (may not be obvious at first that it actually works, but at least the intent is entirely clear :)

**Comment by bennuttall** [comment]
@waveform80 I've pushed a commit which adds the latest 5 blog posts to the homepage (all client-side) in 9a705c11e7e8f0953d1d1d060aef6d85f798f156 - can you merge it and push your branch when you get chance?

**Comment by bennuttall** [comment]
Oh, example here: http://treetrunks.piwheels.org/

**Pull request closed without merging**